Add carto-package.json

rafatower opened the pull request:
Part of https://github.com/CartoDB/onpremises/issues/574

@Algunenano could you please take a look?

Commit:
Add carto-package.json

Algunenano commented on carto-package.json:
```diff
@@ -0,0 +1,17 @@
+    "name": "postgis",
+    "current_version": {
+        "requires": {
+            "postgres": ">=10.1.1"
```
Technically this release (2.4) only requires Postgresql 9.4.

rafatower replied:
I know, you should look into the other issue to have a little bit more context.

The question here is whether our on-premise *dependency system* will support 9.4, not what our fork can actually support.

skgsergio replied:
Well, I think that we should stick not to the versions we have but represent the reality of the piece, I'd put >=9.4

Algunenano commented on carto-package.json:
```diff
@@ -0,0 +1,17 @@
+            "postgres": ">=10.1.1"
+        },
+        "_comment_on_works_with": "Added here versions w/ proper PARALLEL support",
+        "works_with": {
```
I don't understand this section. Shouldn't this be in the other packages requiring this one? I mean, top - bottom, not bottom - up. I'd be madness keeping this up to date.

Even if the final decision it's doing it this way, it would be lacking Windshaft and Mapnik at least (both use Postgis at certain point).

rafatower replied:
Adding @lekum to comment on this.

To my understanding, this is a "soft dependency". Kind of a "recommends" for choosing component versions.

skgsergio replied:
In this case "postgis" doesn't "works with" dataservices or crackshaft there is no postgis-dataservices combo required. But it is in the reverse way, cranckshaft requires (not works with) posgis. Not sure if you get me 🤔

Algunenano replied:
Seeing https://github.com/CartoDB/odbc_fdw/pull/77#discussion_r209938867 I see that we are understanding `works_with` differently. I understood that it's kind of a soft dependency: for example Builder requires certain Windshaft-cartodb or the sql-api versions when running, but it's not a build or even runtime requirement, so it's declared as  `works_with`.
OTOH, Postgis doesn't require crankshaft at all to run, so it shouldn't declare it but the other way around it's true, so it should be declared only on crankshaft carto-package.json.

Am I understanding this incorrectly? I see that here you are declaring all Carto packages Postgis has any relation with, which means that if any package declares `works_with` Postgis, then Postgis needs an equivalent entry here. That would be a huge pain for both Postgis and specially Postgresql since everything `works_with` it :sweat_smile:

skgsergio replied:
Yes, @Algunenano that is. "works_with" is a thing that we are going to use rarely, but the places where it should be used are clear. For example, a CARTO feature introduced in builder requires a version of windshaft and sql-api (or a feature of windshaft requires a version of sql-api and builder) this is the case where "works_with" enters the game, is not a build or even runtime requirement but a feature in order to be fully enabled requires a set of versions.

rafatower replied:
"Reverse soft dependency" is the working definition I used in my head.

Algunenano replied:
> "Reverse soft dependency" is the working definition I used in my head.

I'm using literally the opposite definition :smile: 

> Yes, @Algunenano that is. "works_with" is a thing that we are going to use rarely, but the places where it should be used are clear. For example, a CARTO feature introduced in builder requires a version of windshaft and sql-api (or a feature of windshaft requires a version of sql-api and builder) this is the case where "works_with" enters the game, is not a build or even runtime requirement but a feature in order to be fully enabled requires a set of versions.

So, this means that we should remove the `works_with` from this PR, right? And I guess it's the same for the other ones (odbc_fdw, crankshaft)

Algunenano commented on carto-package.json:
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "postgis",
+    "current_version": {
+        "requires": {
```
I've seen some packages requiring `gdal` in the original issue, so Postgis also should [require](https://trac.osgeo.org/postgis/wiki/UsersWikiPostgreSQLPostGIS):
- gdal > 1.10 (for Raster). Recommended 2.2.
- geos > 3.4. Recommended 3.6 for now (3.7 soon)
- proj.4 . Not sure about minimum or recommended release. The newest the better
- json-c (GeoJSON). Not sure about minimum or recommended release. The newest the better
- protobuf-c (for MVT). Not sure about minimum or recommended release. The newest the better
- libxml2 (for GeomFromGML and GeomFromKML): Not sure about minimum or recommended release. The newest the better

In all cases the package should be compatible with any newer stable release (gdal 2.3, geos 3.7, proj.4 5.1 (great naming), etc) so `>` should be used.

rafatower replied:
The intent of the `carto-package.json` is to specify "CARTO components", face to on-premise build and packaging, and their dependencies among them. This component is considered a leaf of the dependency tree (graph?) as defined by Infrastructure.

I believe the dependencies of postgis itself are managed in some other way (omnibus?).


Adding @skgsergio in case he wants to add/clarify anything.

skgsergio replied:
No worries, is good to have it defined if you feel is not too much work to maintain them (for example @javitonino is doing it for cartodb https://github.com/CartoDB/cartodb/pull/14232/files). We are not going, at least now, to parametrize those dependencies but doesn't hurt to have them there (in fact is good for checking if we are using old/not recommended versions, don't think only in the automatic dep resolution).

rafatower replied:
OK, I'll add them

Commit:
Add some build/link/run requirements

as per the comment
https://github.com/CartoDB/postgis/pull/26/files#r209916760

As minimum versions I used the ones in ubuntu xenial package system,
which are the ones used in cloud builds. Just to have a starting point
that is extensively tested.

Commit:
Remove the entries in `works_with`

Algunenano reviewed (approved):
LGTM
